data/org.freedesktop.portal.DynamicLauncher.xml

```diff
@@ -0,0 +1,223 @@
+              The URL of the web app being installed. This option only makes sense
+              when the launcher is for a web app.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>editable-name b</term>
+            <listitem><para>
+              If true, the user will be able to edit the name of the launcher.
+              Defaults to true.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>editable-icon b</term>
+            <listitem><para>
+              If true, the user will be able to edit the icon of the launcher,
+              if the implementation supports this. Defaults to false.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+
+        The following results get returned via the #org.freedesktop.portal.Request::Response signal:
+        <variablelist>
+          <varlistentry>
+            <term>name s</term>
+            <listitem><para>
+              The name chosen by the user for the launcher.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>token s</term>
+            <listitem><para>
+              Token that can be passed to a subsequent org.freedesktop.portal.DynamicLauncher.Install() call to
+              complete the installation without another dialog.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+    -->
+    <method name="PrepareInstall">
+      <annotation name="org.gtk.GDBus.C.UnixFD" value="true"/>
+      <arg type="s" name="parent_window" direction="in"/>
+      <arg type="s" name="name" direction="in"/>
+      <arg type="h" name="icon_fd" direction="in"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="o" name="handle" direction="out"/>
+    </method>
+    <!--
+        RequestInstallToken:
+        @name: The name that will be used in the desktop file
+        @icon_fd: An open file descriptor for the icon file
+        @options: Vardict with optional further information
+        @token: the token to be used with the Install() method, or the empty
+          string if there was an error
+
+        This method is intended for use only by trusted system components and
+        will fail if used by a sandboxed process. The token returned by this
+        method can be used to avoid the need for a confirmation dialog; the
+        token is passed to the Install() method just as if it were acquired
+        via the PrepareInstall() method.
+
+        The @options vardict currently has no supported entries.
+    -->
+    <method name="RequestInstallToken">
+      <annotation name="org.gtk.GDBus.C.UnixFD" value="true"/>
+      <arg type="s" name="name" direction="in"/>
+      <arg type="h" name="icon_fd" direction="in"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="s" name="token" direction="out"/>
+    </method>
```

Comment:
I find this separation of `PrepareInstall()` and `RequestInstallToken()` not ideal. The portal daemon can detect if an app is sandboxed or not, and therefore can decide when not to ask implementations to show a dialog. If the application is unsandboxed, it would just accept the all the data passed through `PrepareInstall()` unmodified.

Reply:
I don't think that pattern works because I want to use the portal UI in both sandboxed and unsandboxed Epiphany, so we don't have to maintain two code paths there. So the split between `PrepareInstall` and `RequestInstallToken` lets the app choose whether it wants UI to be presented, and the UI also provides editing the name (and potentially the icon eventually) so it is useful.

Reply:
Sorry, I'm still not sold on this approach. The reason I don't think it's a good idea is because portal design generally avoids delegating this kind of decision to apps. x-d-p knows if an app is sandboxed or not, and can make decisions based on that; apps, on the other hand, could call the wrong D-Bus method for a variety of reasons (broken sandbox detection, evil intents, etc.)

I would like to suggest an alternative approach: pass a hint to whether the requesting app wants a dialog or not, as part of the `options` vardict. I imagine it would be something like this:

```
interaction_mode → uint32

The suggested interaction mode. Supported values are:

  * 0: Default. Portal implementations decide how to present this request. This is the default.
  * 1: Interactive. Present a dialog even when not necessary.
```

In this case ^^^, I don't think we need a `SupportedInteractionModes` property, because portal implementations *must* implement all modes.

(Naming TBD; if you can think of better terminology, please let me know.)

Reply:
> Sorry, I'm still not sold on this approach. The reason I don't think it's a good idea is because portal design generally avoids delegating this kind of decision to apps. x-d-p knows if an app is sandboxed or not, and can make decisions based on that; apps, on the other hand, could call the wrong D-Bus method for a variety of reasons (broken sandbox detection, evil intents, etc.)

Sorry I don't see how this is a reason we should move the functionality of `RequestInstallToken()` into `PrepareInstall()`. It is great for x-d-p to make decisions based on whether an app is sandboxed or not, but in this case we do need a way for the app to either install with a dialog or install without, and I think these flows are different enough that it makes sense for them to have different methods. A malicious app that's sandboxed is going to get a dialog regardless of which choice we make here, so I don't understand the point of bringing that up.

> I would like to suggest an alternative approach: pass a hint to whether the requesting app wants a dialog or not, as part of the options vardict. I imagine it would be something like this:
> 
> interaction_mode → uint32
> 
> The suggested interaction mode. Supported values are:
> 
>   * 0: Default. Portal implementations decide how to present this request. This is the default.
>   * 1: Interactive. Present a dialog even when not necessary.

I don't think this makes sense, because we need to support all three of (1) sandboxed Ephy install w/ dialog, (2) unsandboxed Ephy install w/ dialog, and (3) Software install w/o dialog. In your scheme for (3) Software would not be able to specify that there needs to be no interaction (since it has its own UI) and would have to rely on the portal deciding not to show UI, which seems more fragile. It could possibly work as a tri-state, where there's a "Force no interaction" option, but this just seems like over-engineering when we already have a working solution.

Here are the reasons I think the split between the two methods makes sense:

- Having `PrepareInstall` default to not showing a dialog if the app is unsandboxed and the proposed "Force interaction" option isn't set (I think that's what you're suggesting) doesn't make a lot of sense since the user wouldn't have a chance to change the name of the launcher before it gets created (if the `editable-name` option is set). For what it's worth Chrome shows a confirmation dialog when installing a PWA even without letting the user change the launcher name, so maybe that's more evidence that it's the right UX.
- The `RequestInstallToken` method implementation has a check at the top to error out if the caller is sandboxed, and it's easy to see that this is correct when reading the code, whereas a similar check in `PrepareInstall` would have to be conditional on the `Force no interaction` option being set and therefore a bit harder to check for correctness.
- There's a lot of complexity in `PrepareInstall` that is not relevant when the caller does not want a dialog: the `parent_window` argument, the use of the `org.freedesktop.portal.Request` pattern, and all of the keys in the `options` vardict. I don't see how it makes sense to force clients to deal with all that when they don't need it (and to force x-d-p handle those things being omitted correctly).
- For what it's worth there's already precedent for methods that are only for unsandboxed callers: e.g. `org.freedesktop.portal.Device.AccessDevice()`

I do appreciate your time reviewing this and I don't intend to be combative.

Reply:
If you'd like to take this to a court of appeals we could ask a third party for an opinion :laughing:

Reply:
I'd like to have third and fourth party opinions here, for sure :slightly_smiling_face:

Reply:
Maybe @pwithnall since he's been looking at the Software side of this already?

Reply:
For what it's worth the `PrepareInstall()` method here is modeled after the `PreparePrint()` method in the Print portal, which creates a dialog regardless of whether or not the caller is sandboxed.

Reply:
As noted [below](https://github.com/flatpak/xdg-desktop-portal/pull/696#issuecomment-1051090694), we came to a resolution on this discussion.